heroku/buildpacks-go
# Support ARM64

This is a WIP for CI/integration testing, and should not be reviewed independently as it will be rewritten.

It integrates multiple branches currently in review:

* https://github.com/heroku/buildpacks-go/pull/229
* https://github.com/heroku/buildpacks-go/pull/230 (the base for this PR)
* Merges in https://github.com/heroku/buildpacks-go/pull/231 for integration testing
* Finally, https://github.com/heroku/buildpacks-go/pull/232/commits/e22de205768b19f5db5c9fcd485e0acfdeb49d81, on this branch, adds the remaining buildpack specific changes required for multi-arch support (e.g. resolving, caching, restoring and installing the correct dependencies), and support for building the required CNB binaries for `x86_64` + `aarch64` by updating `buildpack.toml`.

For testing the buildpack, an ARM64 compatible builder can be created locally on (using the newly released `heroku-24` base images (tested on an m1 Mac with Docker's `containerd` feature enabled), and a basic builder configuration file) - e.g. something like this:

`builder-24/builder.toml`:
```
description = "Heroku-24 (Ubuntu 24.04) builder sans any buildpacks"

[stack]
id = "heroku-24"
build-image = "heroku/heroku:24-build.nightly"
run-image = "heroku/heroku:24.nightly"

[lifecycle]
version = "0.19.0"
```

Create the builder:

    pack builder create heroku-24:builder-foo --config ./builder-24/builder.toml

Compile and package the CNBs:

    cargo libcnb package

    cargo libcnb package --target aarch64-unknown-linux-musl

To test that the `x86_64` version still work as intended, we need to use the old builder (as`heroku-24:builder-foo` won't work for multi-arch locally):
```
pack build go-amd64-app \
--buildpack packaged/x86_64-unknown-linux-musl/debug/heroku_go \
--path buildpacks/go/tests/fixtures/basic_http_122 --builder heroku/builder:22
```

To test `aarch64`:

```
pack build go-arm64-app \
--buildpack packaged/aarch64-unknown-linux-musl/debug/heroku_go \
--path buildpacks/go/tests/fixtures/basic_http_122 --builder heroku-24:builder-foo
```

Running the commands twice should reuse cached layers as usual. Another app can be used to verify cache invalidation, e.g.:

```
pack build go-arm64-app \
--buildpack packaged/aarch64-unknown-linux-musl/debug/heroku_go \
--path buildpacks/go/tests/fixtures/basic_http_119 --builder heroku-24:builder-foo
```
